File: modules/EphysLink.jl/src/session.jl
# The Session object in Julia. Mirrors modules/ephyslink/session.py field for field.
#
# One mutable struct rather than a type hierarchy: `kind` says whether it came from continuous
# data ("OE") or a spike sorter ("KS"), and the accessors below check it. A Julia subtype per
# kind would buy nothing here — the file format is identical, and dispatch on a string field
# is enough for two cases.

using Dates

"""
    Session

One recording plus everything an analysis pipeline has attached to it.

Fields:

- `kind`          `"OE"` (continuous) or `"KS"` (spike-sorted)
- `id`            session identifier
- `experiment`    free text
- `fs_hz`         sampling rate
- `source`        where the data originally came from
- `arrays`        name → array. Axis order matches Python unless `native` is true
- `dims`          name → axis names, describing `arrays` as they currently are
- `array_meta`    name → units / description
- `tables`        name → (column name → vector)
- `table_columns` name → column order, which a Dict does not preserve
- `events`        name → sample indices
- `meta`          arbitrary metadata
- `history`       one entry per pipeline step
- `native`        true if arrays are in HDF5-native (reversed) orientation
"""
mutable struct Session
    kind::String
    id::String
    experiment::String
    fs_hz::Float64
    source::String

    arrays::Dict{String,Any}
    dims::Dict{String,Vector{String}}
    array_meta::Dict{String,Dict{String,String}}

    tables::Dict{String,Dict{String,Any}}
    table_columns::Dict{String,Vector{String}}

    events::Dict{String,Vector{Int64}}

    meta::Any
    history::Any

    native::Bool
end

"""
    Session(kind, id, fs_hz; experiment="", source="")

An empty session, for building one up in Julia rather than loading it.
"""
Session(kind::AbstractString, id::AbstractString, fs_hz::Real;
        experiment::AbstractString = "", source::AbstractString = "") =
    Session(String(kind), String(id), String(experiment), Float64(fs_hz), String(source),
            Dict{String,Any}(), Dict{String,Vector{String}}(),
            Dict{String,Dict{String,String}}(),
            Dict{String,Dict{String,Any}}(), Dict{String,Vector{String}}(),
            Dict{String,Vector{Int64}}(),
            Dict{String,Any}(), Any[], false)


# ---- attaching things -------------------------------------------------------

"""
    add_array!(s, name, A, dims; units=nothing, description=nothing)

Attach an array. `dims` is required — an unlabelled axis is how a session gets silently
transposed between languages, so the format does not allow one.
"""
function add_array!(s::Session, name::AbstractString, A::AbstractArray,
                    dims::AbstractVector{<:AbstractString};
                    units = nothing, description = nothing)
    length(dims) == ndims(A) || error(
        "array \"$(name)\" has $(ndims(A)) axes but $(length(dims)) names were given: $(dims)"
    )
    s.arrays[String(name)] = A
    s.dims[String(name)] = String.(collect(dims))
    extra = Dict{String,String}()
    units === nothing || (extra["units"] = String(units))
    description === nothing || (extra["description"] = String(description))
    isempty(extra) || (s.array_meta[String(name)] = extra)
    return s
end

"""
    add_table!(s, name, columns; order=collect(keys(columns)))

Attach a table as a `Dict` of column name → vector. `order` fixes the column order, which a
Dict does not preserve and which Python's DataFrame will otherwise show scrambled.
"""
function add_table!(s::Session, name::AbstractString, columns::AbstractDict;
                    order::AbstractVector{<:AbstractString} = collect(keys(columns)))
    lengths = unique(length(v) for v in values(columns))
    length(lengths) == 1 || error("table \"$(name)\" has columns of differing length: $(lengths)")
    s.tables[String(name)] = Dict{String,Any}(String(k) => v for (k, v) in columns)
    s.table_columns[String(name)] = String.(collect(order))
    return s
end

add_events!(s::Session, name::AbstractString, samples) =
    (s.events[String(name)] = Vector{Int64}(samples); s)

"""
    log!(s, step; params...)

Record a pipeline step. Saved with the session, so a file explains itself.
"""
function log!(s::Session, step::AbstractString; params...)
    push!(s.history, Dict{String,Any}(
        "step" => String(step),
        "params" => Dict{String,Any}(String(k) => v for (k, v) in params),
        "time" => string(Dates.now(Dates.UTC)),
    ))
    return s
end


# ---- reading things ---------------------------------------------------------

function array(s::Session, name::AbstractString)
    haskey(s.arrays, name) ||
        error("no array \"$(name)\". Available: $(sort(collect(keys(s.arrays))))")
    return s.arrays[name]
end

function table(s::Session, name::AbstractString)
    haskey(s.tables, name) ||
        error("no table \"$(name)\". Available: $(sort(collect(keys(s.tables))))")
    return s.tables[name]
end

"""
    axis(s, array_name, axis_name) -> Int

Position of a named axis, **1-based** as Julia expects. Use this instead of hard-coding an
index: it is the difference between code that survives a transposed array and code that does
not.
"""
function axis(s::Session, array_name::AbstractString, axis_name::AbstractString)
    names = get(s.dims, array_name, String[])
    index = findfirst(==(axis_name), names)
    index === nothing &&
        error("array \"$(array_name)\" has axes $(names), not \"$(axis_name)\"")
    return index
end

n_along(s::Session, array_name::AbstractString, axis_name::AbstractString) =
    size(array(s, array_name), axis(s, array_name, axis_name))

"""
    seconds(s, samples)

Sample indices to seconds. The one place the sampling rate is divided by.
"""
seconds(s::Session, samples) = samples ./ s.fs_hz


# ---- kind-specific convenience ---------------------------------------------

_require(s::Session, kind::AbstractString, what::AbstractString) =
    s.kind == kind || error("$(what) is only defined for kind=\"$(kind)\" sessions; this is \"$(s.kind)\"")

"""
    slice_along(s, array_name, axis_name, i) -> SubArray

A zero-copy view with the named axis fixed at `i`, whatever position that axis occupies.

This is what makes code layout-independent. `slice_along(s, "continuous", "channel", 3)` is
one channel's samples in both `native` and matched mode, with no `if` at the call site and no
hard-coded index.

It is a *view*: nothing is copied, and writing through it writes into the session.

Speed, stated because it is the whole reason `native` defaults to true: in native mode the
returned view of a channel is a **contiguous column**. In matched mode it is a strided row, so
it is correct but slow. Same code, same answer, different memory behaviour.
"""
function slice_along(s::Session, array_name::AbstractString,
                     axis_name::AbstractString, i::Integer)
    A = array(s, array_name)
    k = axis(s, array_name, axis_name)
    return view(A, ntuple(d -> d == k ? i : Colon(), ndims(A))...)
end

continuous(s::Session) = (_require(s, "OE", "continuous"); array(s, "continuous"))

"""
    n_channels(s; array="continuous")
    channel(s, i; array="continuous")
    each_channel(s; array="continuous")

One channel's data, by number, regardless of layout mode.

    for trace in each_channel(s)
        # `trace` is a view of one channel's samples, contiguous under the default layout
    end
"""
n_channels(s::Session; array::AbstractString = "continuous") = n_along(s, array, "channel")
channel(s::Session, i::Integer; array::AbstractString = "continuous") =
    slice_along(s, array, "channel", i)
each_channel(s::Session; array::AbstractString = "continuous") =
    (channel(s, i; array = array) for i in 1:n_channels(s; array = array))

n_samples(s::Session; array::AbstractString = "continuous") = n_along(s, array, "sample")

"""
    n_epochs(s; array="epochs")
    epoch(s, i; array="epochs")
    each_epoch(s; array="epochs")

One epoch, by number. Same idea as `channel`: the axis is found by name, so the code does not
change when the array is stored the other way round.
"""
n_epochs(s::Session; array::AbstractString = "epochs") = n_along(s, array, "epoch")
epoch(s::Session, i::Integer; array::AbstractString = "epochs") =
    slice_along(s, array, "epoch", i)
each_epoch(s::Session; array::AbstractString = "epochs") =
    (epoch(s, i; array = array) for i in 1:n_epochs(s; array = array))

spike_times(s::Session) = (_require(s, "KS", "spike_times"); array(s, "spike_times"))
spike_clusters(s::Session) = (_require(s, "KS", "spike_clusters"); array(s, "spike_clusters"))
n_spikes(s::Session) = length(spike_times(s))

"""
    spikes_of(s, cluster)

One cluster's spike times, in samples. Linear in the number of spikes — for repeated access
across many clusters, bucket once instead.
"""
spikes_of(s::Session, cluster::Integer) = spike_times(s)[spike_clusters(s) .== cluster]

function duration_s(s::Session)
    if s.kind == "OE"
        return n_samples(s) / s.fs_hz
    else
        t = spike_times(s)
        return isempty(t) ? 0.0 : (maximum(t) - minimum(t)) / s.fs_hz
    end
end


# ---- validation and display -------------------------------------------------

function validate(s::Session)
    for (name, A) in s.arrays
        labels = get(s.dims, name, String[])
        length(labels) == ndims(A) || error(
            "array \"$(name)\" has $(ndims(A)) axes but dims $(labels). Every array must " *
            "name its axes — an unlabelled axis is how a session gets silently transposed " *
            "between languages."
        )
    end
    s.fs_hz > 0 || error("fs_hz is $(s.fs_hz)")
    if s.kind == "KS" && haskey(s.arrays, "spike_times")
        t = s.arrays["spike_times"]
        issorted(t) || error("spike_times is not sorted; every searchsorted downstream assumes it is")
    end
    return s
end

function Base.show(io::IO, s::Session)
    parts = ["Session(\"$(s.id)\"", "kind=$(s.kind)", "fs=$(s.fs_hz) Hz"]
    isempty(s.arrays) || push!(parts, "$(length(s.arrays)) arrays")
    isempty(s.tables) || push!(parts, "$(length(s.tables)) tables")
    isempty(s.events) || push!(parts, "$(length(s.events)) event channels")
    print(io, join(parts, ", "), ")")
end

"""
    overview(s) -> String

Everything in the session, in the order it would be written.

Named `overview` rather than `summary` deliberately: `Base.summary` exists, and a method that
shadows it would break unrelated code in any session that does `using EphysLink`.
"""
function overview(s::Session)
    io = IOBuffer()
    println(io, "Session  ", s.id, "  [", s.kind, "]")
    println(io, "  experiment  ", isempty(s.experiment) ? "—" : s.experiment)
    println(io, "  fs          ", s.fs_hz, " Hz")
    println(io, "  source      ", isempty(s.source) ? "—" : s.source)
    s.native && println(io, "  layout      NATIVE (axes reversed relative to Python)")
    if !isempty(s.arrays)
        println(io, "  arrays")
        for name in sort(collect(keys(s.arrays)))
            A = s.arrays[name]
            println(io, "    ", rpad(name, 20), rpad(string(size(A)), 22), eltype(A),
                    "  [", join(get(s.dims, name, String[]), ", "), "]")
        end
    end
    if !isempty(s.tables)
        println(io, "  tables")
        for name in sort(collect(keys(s.tables)))
            columns = s.table_columns[name]
            println(io, "    ", rpad(name, 20), length(s.tables[name][first(columns)]),
                    " rows × ", length(columns), " columns")
        end
    end
    if !isempty(s.events)
        println(io, "  events")
        for name in sort(collect(keys(s.events)))
            println(io, "    ", rpad(name, 20), length(s.events[name]), " events")
        end
    end
    return String(take!(io))
end

File: modules/ephyslink/session.py
"""
The Session object: one recording, and everything the pipeline attaches to it.

The point of this class is that a pipeline stage should not produce a new file. It should
take a session, add its output to it, and hand the same object on. Whatever is in `arrays`,
`tables` and `events` at the end is what `save()` writes, and `load()` gives back — in Python
or in Julia, with the same axis order and the same names.

    session = load_kilosort("…/2026-04-14_11-46-55")
    session.add_table("units", unit_table)
    session.add_array("psth", psth_matrix, dims=["unit", "bin"])
    session.log("responder test", alpha=0.05)
    session.save("…/analysis/2026-04-14.h5")

Two typed views over one file format:

* `SessionOE` — continuous traces. LFP, MEP, EEG. The data is `(channel, sample)`, and after
  epoching `(channel, sample, epoch)`.
* `SessionKS` — spike-sorted output. Spike times, clusters, templates, and a cluster table.

They differ only in convenience accessors and in what `validate()` insists on; the file they
write is the same format, distinguished by the `kind` attribute. Adding a third kind means
adding a subclass, not a new format.

**Axis order is never inferred.** Every array is stored with a `dims` attribute naming its
axes. `session.axis("continuous", "channel")` gives the position of an axis by name, so
analysis code does not have to hard-code 0 or 1 and does not break when an array is
transposed upstream.
"""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterable

import numpy as np
import pandas as pd

from . import format as fmt


class Session:
    """Base class. Use `SessionOE` or `SessionKS`; this holds what they share."""

    kind = "base"

    def __init__(
        self,
        id: str,
        fs_hz: float,
        experiment: str = "",
        source: str | Path = "",
        arrays: dict[str, np.ndarray] | None = None,
        dims: dict[str, list[str]] | None = None,
        tables: dict[str, pd.DataFrame] | None = None,
        events: dict[str, np.ndarray] | None = None,
        meta: dict[str, Any] | None = None,
        history: list[dict] | None = None,
        array_meta: dict[str, dict] | None = None,
    ):
        self.id = str(id)
        self.fs_hz = float(fs_hz)
        self.experiment = str(experiment or "")
        self.source = str(source or "")
        self.arrays: dict[str, np.ndarray] = dict(arrays or {})
        self.dims: dict[str, list[str]] = dict(dims or {})
        self.tables: dict[str, pd.DataFrame] = dict(tables or {})
        self.events: dict[str, np.ndarray] = dict(events or {})
        self.meta: dict[str, Any] = dict(meta or {})
        self.history: list[dict] = list(history or [])
        self.array_meta: dict[str, dict] = dict(array_meta or {})

    # ---- results ----------------------------------------------------------
    #
    # `results` is a free-form dict for the session-level numbers an analysis produces —
    # counts, medians, test outcomes. It is stored inside `meta`, so it round-trips through
    # the file and into Julia with no extra machinery and no format change.
    #
    # Derived *tables* do not go here. They go in `tables` like any other table; what makes
    # them results is that they were not there when the session was loaded. See
    # `mark_source()` and `export_results()`.

    @property
    def results(self) -> dict[str, Any]:
        """Session-level numbers, accumulated as the analysis runs.

            session.results["n_clean_units"] = 45
            session.results["median_ppc_ctx_20hz"] = 0.0277
        """
        return self.meta.setdefault("results", {})

    def mark_source(self) -> "Session":
        """Record what is in the session *now* as raw input rather than analysis output.

        Called by the loaders once they have finished. Everything added afterwards is, by
        definition, something the analysis produced — which is what lets `export_results`
        strip the raw data without being told what the raw data was.
        """
        self.meta["source_arrays"] = sorted(self.arrays)
        self.meta["source_tables"] = sorted(self.tables)
        self.meta["source_events"] = sorted(self.events)
        return self

    def derived_tables(self) -> list[str]:
        """Tables added since `mark_source()`. Everything if the source was never marked."""
        source = set(self.meta.get("source_tables", []))
        return [name for name in self.tables if name not in source]

    def derived_arrays(self) -> list[str]:
        source = set(self.meta.get("source_arrays", []))
        return [name for name in self.arrays if name not in source]

    def export_results(
        self,
        path: str | Path,
        include_arrays: list[str] | None = None,
        drop_meta: tuple[str, ...] = ("kilosort_params", "source_arrays",
                                      "source_tables", "source_events"),
    ) -> Path:
        """Write a compact results-only session: no spikes, no traces, no templates.

        Keeps every table the analysis added, at full per-row resolution, plus `results`,
        the metadata and the full history. A comparison across sessions then needs only these
        files — for the EXELU probe session that is 859 KB against 44.6 MB with the raw data in
        it, a factor of 52.

        Arrays are excluded by default: they are the bulky intermediates (PSTH matrices,
        cycle histograms) and are re-derivable from the raw session. Name the ones you want
        in `include_arrays`.

        The result is an ordinary ephyslink file with `kind="RESULTS"`, so it loads with the
        same `Session.load` and the same Julia reader.
        """
        derived = self.derived_tables()
        if not derived and not self.results:
            raise ValueError(
                f"session '{self.id}' has no derived tables and an empty results dict — "
                "there is nothing to export. Attach analysis output with add_table() and "
                "session.results[...] first."
            )

        meta = {k: v for k, v in self.meta.items() if k not in drop_meta}
        meta["exported_from"] = self.kind
        meta["source_tables_dropped"] = sorted(set(self.tables) - set(derived))

        export = SessionResults(
            id=self.id, fs_hz=self.fs_hz, experiment=self.experiment, source=self.source,
            tables={name: self.tables[name] for name in derived},
            meta=meta, history=list(self.history),
        )
        for name in include_arrays or []:
            export.add_array(name, self.arrays[name], dims=self.dims[name])

        export.log("export_results", tables=derived, arrays=list(include_arrays or []),
                   n_results=len(self.results))
        return fmt.write_session(path, export)

    # ---- attaching things -------------------------------------------------

    def add_array(
        self, name: str, values, dims: Iterable[str],
        units: str | None = None, description: str | None = None,
    ) -> "Session":
        """Attach an array. `dims` is required — an unlabelled axis is how sessions get
        silently transposed, so the format does not allow one."""
        values = np.asarray(values)
        dims = [str(d) for d in dims]
        if len(dims) != values.ndim:
            raise ValueError(
                f"array '{name}' has {values.ndim} axes but {len(dims)} names were given: {dims}"
            )
        if values.dtype.kind not in "fiub":
            # Caught here rather than at save time, so the error points at the line that made
            # the mistake. Strings and object arrays have no portable HDF5 representation —
            # h5py and HDF5.jl disagree — so they belong in a table column.
            raise TypeError(
                f"array '{name}' has dtype {values.dtype}; arrays must be numeric or boolean. "
                "Strings and mixed types belong in a table: session.add_table(...)."
            )
        self.arrays[name] = values
        self.dims[name] = dims
        extra = {k: v for k, v in (("units", units), ("description", description)) if v}
        if extra:
            self.array_meta[name] = extra
        return self

    def add_table(self, name: str, frame: pd.DataFrame) -> "Session":
        self.tables[name] = frame.reset_index(drop=True)
        return self

    def add_events(self, name: str, samples) -> "Session":
        self.events[name] = np.asarray(samples, dtype=np.int64).ravel()
        return self

    def log(self, step: str, **params) -> "Session":
        """Record a pipeline step. Saved with the session, so a file explains itself."""
        self.history.append({
            "step": str(step),
            "params": params,
            "time": datetime.now(timezone.utc).isoformat(),
        })
        return self

    # ---- reading things ---------------------------------------------------

    def array(self, name: str) -> np.ndarray:
        if name not in self.arrays:
            raise KeyError(f"no array '{name}'. Available: {sorted(self.arrays)}")
        return self.arrays[name]

    def table(self, name: str) -> pd.DataFrame:
        if name not in self.tables:
            raise KeyError(f"no table '{name}'. Available: {sorted(self.tables)}")
        return self.tables[name]

    def axis(self, array_name: str, axis_name: str) -> int:
        """Position of a named axis. Use this instead of hard-coding 0 or 1."""
        names = self.dims.get(array_name, [])
        if axis_name not in names:
            raise KeyError(f"array '{array_name}' has axes {names}, not '{axis_name}'")
        return names.index(axis_name)

    def n_along(self, array_name: str, axis_name: str) -> int:
        """Length of a named axis."""
        return int(self.arrays[array_name].shape[self.axis(array_name, axis_name)])

    def seconds(self, samples) -> np.ndarray:
        """Sample indices to seconds. The one place the sampling rate is divided by."""
        return np.asarray(samples, dtype=np.float64) / self.fs_hz

    # ---- input / output ---------------------------------------------------

    def save(self, path: str | Path, compression: str | None = "gzip") -> Path:
        self.validate()
        return fmt.write_session(path, self, compression=compression)

    @staticmethod
    def load(path: str | Path) -> "Session":
        """Load any ephyslink file, returning the subclass its `kind` attribute names."""
        data = fmt.read_session_dict(path)
        cls = {"OE": SessionOE, "KS": SessionKS, "RESULTS": SessionResults}.get(
            data["kind"], Session)
        session = cls(
            id=data["id"], fs_hz=data["fs_hz"], experiment=data["experiment"],
            source=data["source"], arrays=data["arrays"], dims=data["dims"],
            tables=data["tables"], events=data["events"], meta=data["meta"],
            history=data["history"], array_meta=data["array_meta"],
        )
        session.meta.setdefault("created", data["created"])
        return session

    def validate(self) -> None:
        """Checks that would otherwise fail silently, or fail later in Julia."""
        for name, values in self.arrays.items():
            declared = self.dims.get(name)
            if not declared or len(declared) != values.ndim:
                raise ValueError(
                    f"array '{name}' has shape {values.shape} but dims {declared}. Every array "
                    "must name its axes — an unlabelled axis is how a session gets silently "
                    "transposed between languages."
                )
        if self.fs_hz <= 0:
            raise ValueError(f"fs_hz is {self.fs_hz}")

    # ---- niceties ---------------------------------------------------------

    def __repr__(self) -> str:
        parts = [f"{type(self).__name__}('{self.id}'", f"fs={self.fs_hz:g} Hz"]
        if self.arrays:
            parts.append(f"{len(self.arrays)} arrays")
        if self.tables:
            parts.append(f"{len(self.tables)} tables")
        if self.events:
            parts.append(f"{len(self.events)} event channels")
        return ", ".join(parts) + ")"

    def summary(self) -> str:
        """Everything in the session, in the order it would be written."""
        lines = [f"{type(self).__name__}  {self.id}",
                 f"  experiment  {self.experiment or '—'}",
                 f"  fs          {self.fs_hz:g} Hz",
                 f"  source      {self.source or '—'}"]
        if self.arrays:
            lines.append("  arrays")
            for name, values in self.arrays.items():
                dims = ", ".join(self.dims.get(name, []))
                lines.append(f"    {name:<20} {str(values.shape):<22} {values.dtype}  [{dims}]")
        if self.tables:
            lines.append("  tables")
            for name, frame in self.tables.items():
                lines.append(f"    {name:<20} {len(frame)} rows × {len(frame.columns)} columns")
        if self.events:
            lines.append("  events")
            for name, samples in self.events.items():
                lines.append(f"    {name:<20} {samples.size} events")
        if self.meta:
            lines.append(f"  meta        {', '.join(sorted(self.meta))}")
        if self.history:
            lines.append("  history")
            for entry in self.history:
                lines.append(f"    {entry['time'][:19]}  {entry['step']}")
        return "\n".join(lines)


# =============================================================================

class SessionResults(Session):
    """What an analysis produced, with the raw data stripped out.

    Produced by `Session.export_results()`, never by a loader. Holds the derived tables at
    full per-row resolution, the `results` dict of session-level numbers, the metadata and the
    history — everything a cross-session comparison needs and nothing it does not.

    It is an ordinary ephyslink file, so Julia reads it with the same code.
    """

    kind = "RESULTS"

    def validate(self) -> None:
        super().validate()
        if not self.tables and not self.results:
            raise ValueError(f"results session '{self.id}' is empty")


class SessionOE(Session):
    """Continuous data — LFP, MEP, EEG. `continuous` is (channel, sample)."""

    kind = "OE"

    @property
    def continuous(self) -> np.ndarray:
        return self.array("continuous")

    @property
    def n_channels(self) -> int:
        return self.n_along("continuous", "channel")

    @property
    def n_samples(self) -> int:
        return self.n_along("continuous", "sample")

    @property
    def duration_s(self) -> float:
        return self.n_samples / self.fs_hz

    @property
    def channel_names(self) -> list[str]:
        return list(self.meta.get("channel_names", []))

    @property
    def times_s(self) -> np.ndarray:
        """Time axis aligned to `continuous`, in seconds."""
        return np.arange(self.n_samples, dtype=np.float64) / self.fs_hz

    def validate(self) -> None:
        super().validate()
        if "continuous" in self.arrays:
            names = self.dims["continuous"]
            if names[:2] != ["channel", "sample"]:
                raise ValueError(
                    f"SessionOE 'continuous' must be (channel, sample); its dims are {names}. "
                    "Transpose it before attaching rather than relabelling the axes."
                )


class SessionKS(Session):
    """Spike-sorted output. Spikes are sample indices; divide by `fs_hz` once, at the edge."""

    kind = "KS"

    @property
    def spike_times(self) -> np.ndarray:
        return self.array("spike_times")

    @property
    def spike_clusters(self) -> np.ndarray:
        return self.array("spike_clusters")

    @property
    def clusters(self) -> pd.DataFrame:
        return self.table("clusters")

    @property
    def n_spikes(self) -> int:
        return int(self.spike_times.size)

    @property
    def duration_s(self) -> float:
        times = self.spike_times
        return float(times.max() - times.min()) / self.fs_hz if times.size else 0.0

    def spikes_of(self, cluster: int) -> np.ndarray:
        """One cluster's spike times, in samples.

        Linear in the number of spikes. For repeated access across many clusters, bucket once
        with `scripts/spiketrains.py::SpikeStore` instead — this is the convenience path, not
        the fast one.
        """
        return self.spike_times[self.spike_clusters == int(cluster)]

    def validate(self) -> None:
        super().validate()
        n = self.arrays.get("spike_times", np.empty(0)).size
        for name in ("spike_clusters", "spike_positions", "spike_templates"):
            if name in self.arrays and len(self.arrays[name]) != n:
                raise ValueError(
                    f"'{name}' has {len(self.arrays[name])} rows but spike_times has {n}"
                )
        if "spike_times" in self.arrays and not np.all(np.diff(self.spike_times) >= 0):
            raise ValueError("spike_times is not sorted; every searchsorted downstream assumes it is")
